### ISSUE TITLE
Error: Attribute name required

### ISSUE THREAD

**Issue by mrmodolo** [hashicorp/terraform#22307]
Hi!

After upgrading from version 0.12.4 to version 0.12.6 I can no longer ignore a specific attribute in lifecycle!

```
resource "vsphere_virtual_machine" "vm" {
  lifecycle {
    ignore_changes = [
      "boot_delay",
      "guest_id",
      "custom_attributes.103",
    ]
  }
```

### Terraform Version

Terraform v0.12.6

### Expected Behavior

Prior to version 0.12.6 I could ignore changes to specific attributes in "custom_attributes". For example, I could add "custom_attributes.103" to ignore_changes and follow with terraform plan or apply without errors.

### Actual Behavior
With version 0.12.6 if I use (quotes) "custom_attributes.103" in ignore_changes, I get the error "Dot must be followed by attribute name."

If I use custom_attributes.103 without quotes, the error is not displayed but the attribute is not ignored! I even tried the custom_attributes ["103"] syntax, which showed no errors but again, the attribute was not ignored.

### Steps to Reproduce
1. `terraform plan`
2. `terraform apply`

### Additional Context

This behavior started with version 0.12.5! In version 0.12.4 everything works.
If I only use custom_attributes in ignore_changes all attributes are ignored!

### References
https://github.com/terraform-providers/terraform-provider-vsphere/issues/815

**Comment by teamterraform** [comment]
Hello, to better understand this issue & make it easier to reproduce for anyone else would you mind sharing more details? e.g.

-  the full configuration of `resource "vsphere_virtual_machine" "vm"` (or a simpler reproduction case)
 - full output (stdout + stderr) from terraform commands you ran (minus any secrets)
 - [debug log](https://github.com/hashicorp/terraform/blob/master/website/source/docs/internals/debugging.html.md#debugging-terraform) from terraform commands you ran (minus any secrets)

Thanks.

**Comment by mrmodolo** [comment]
Hi!

Thanks for the feedback, I am posting the supplementary information!

First, trying to apply using "custom_attributes.103" in quotes!

```
lifecycle {
    ignore_changes = [
      boot_delay, 
      guest_id,
      "custom_attributes.103"
    ]
  }
```

```
TF_LOG=DEBUG terraform  apply -target=module.production.module.cyberark
2019/08/02 10:12:49 [INFO] Terraform version: 0.12.6  
2019/08/02 10:12:49 [INFO] Go runtime version: go1.12.4
2019/08/02 10:12:49 [INFO] CLI args: []string{"/usr/local/bin/terraform", "apply", "-target=module.production.module.cyberark"}
2019/08/02 10:12:49 [DEBUG] Attempting to open CLI config file: /home/modolo/.terraformrc
2019/08/02 10:12:49 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2019/08/02 10:12:49 [INFO] CLI command args: []string{"apply", "-target=module.production.module.cyberark"}
2019/08/02 10:12:49 [DEBUG] checking for provider in "."
2019/08/02 10:12:49 [DEBUG] checking for provider in "/usr/local/bin"
2019/08/02 10:12:49 [DEBUG] checking for provider in ".terraform/plugins/linux_amd64"
2019/08/02 10:12:49 [DEBUG] found provider "terraform-provider-vsphere_v1.12.0_x4"
2019/08/02 10:12:49 [DEBUG] found valid plugin: "vsphere", "1.12.0", "/home/modolo/terraform-vmware/GLOBOSAT-RJ2/.terraform/plugins/linux_amd64/terraform-provider-vsphere_v1.12.0_x4"
2019/08/02 10:12:49 [DEBUG] checking for provisioner in "."
2019/08/02 10:12:49 [DEBUG] checking for provisioner in "/usr/local/bin"
2019/08/02 10:12:49 [DEBUG] checking for provisioner in ".terraform/plugins/linux_amd64"
2019/08/02 10:12:49 [INFO] backend/local: starting Apply operation

Error: Attribute name required

  on production/cyberark/modules/vmware/main.tf line 57, in resource "vsphere_virtual_machine" "vm":
  57:       "custom_attributes.103"

Dot must be followed by attribute name.


Error: Attribute name required

  on production/cyberark/modules/vmware/main.tf line 57, in resource "vsphere_virtual_machine" "vm":
  57:       "custom_attributes.103"

Dot must be followed by attribute name.
```
Now without using quotes!
TF_LOG=DEBUG terraform  plan -target=module.production.module.cyberark
The output in debug mode is attached!

[module.production.module.cyberark.txt](https://github.com/hashicorp/terraform/files/3461408/module.production.module.cyberark.txt)

```
lifecycle {
    ignore_changes = [
      boot_delay, 
      guest_id,
      custom_attributes.103
    ]
  }
```
```
# **module.production.module.cyberark.module.PFXSRV0000492.module.vmware.vsphere_virtual_machine.vm will be updated in-place**
  ~ resource "vsphere_virtual_machine" "vm" {
        boot_delay                              = 5000
        boot_retry_delay                        = 10000
        boot_retry_enabled                      = false
        change_version                          = "2019-07-31T08:35:12.925206Z"
        cpu_hot_add_enabled                     = false
        cpu_hot_remove_enabled                  = false
        cpu_limit                               = -1
        cpu_performance_counters_enabled        = false
        cpu_reservation                         = 0
        cpu_share_count                         = 4000
        cpu_share_level                         = "normal"
      ~ custom_attributes                       = {
            "1"    = "Marcelo Módolo <modolo@globosat.com.br>"
            "101"  = "Daniel Dezerto Castanha <dcastanha@globosat.com.br>"
            "102"  = "55 21 2145-7237"
          **- "103"  = "28/07/2019 23:06:45" -> null**
            "1601" = "SUM Servers Manual"
            "1701" = "SR940082"
            "301"  = "Solução PAM da CyberArk"
            "302"  = "PSMP Solução PAM da CyberArk"
            "401"  = "TECTT-TI-SEGURANCA"
            "501"  = "04/07/2019"
            "601"  = "55 21 2145-7238"
            "602"  = "Antonio Horta <horta@globosat.com.br>"
        }
```

Following is my main.tf file!

```hcl

locals {
  common_attributes = {
    "${module.attributes.analista_responsavel_id}"            = "${var.analista_responsavel}" 
    "${module.attributes.analista_criacao_id}"                = "${var.analista_criacao}"
    "${module.attributes.data_criacao_id}"                    = "${var.data_criacao}"
    "${module.attributes.chamado_criacao_id}"                 = "${var.chamado_criacao}"
    "${module.attributes.departamento_area_id}"               = "${var.departamento_area}"
    "${module.attributes.descricao_servico_sistema_id}"       = "${var.descricao_servico_sistema}"
    "${module.attributes.gestor_ponto_focal_solicitante_id}"  = "${var.gestor_ponto_focal_solicitante}"
    "${module.attributes.janela_wsus_id}"                     = "${var.janela_wsus}"
    "${module.attributes.justificativa_id}"                   = "${var.justificativa}"
    "${module.attributes.ramal_analista_responsavel_id}"      = "${var.ramal_analista_responsavel}"
    "${module.attributes.ramal_analista_criacao_id}"          = "${var.ramal_analista_criacao}"
  }
}

module "datacenter" {
  source = "../../../../modules/datacenter/"
}

module "attributes" {
  source = "../../../../modules/attributes"
}

module "computer_cluster" {
  source = "../../../../modules/computer_cluster"
}

module "datastore_cluster" {
  source = "../../../../modules/datastore_cluster"
}

module "datacenter_templates" {
  source = "../../../../../GLOBOSAT-RJ/modules/datacenter"
}

data "vsphere_network" "network" {
  name          = "${var.network_interface_name}"
  datacenter_id = "${module.datacenter.id}"
}

data "vsphere_virtual_machine" "template" {
  name          = "${var.template_name}"
  datacenter_id = "${module.datacenter_templates.id}"
}

resource "vsphere_virtual_machine" "vm" {

  lifecycle {
    ignore_changes = [
      boot_delay, 
      guest_id,
      custom_attributes.103
    ]
  }

  wait_for_guest_net_timeout  = 0
  wait_for_guest_net_routable = false
  force_power_off             = true

  folder = "CyberArk"
  
  name                      = "${var.name}"
 
  custom_attributes         = "${merge(local.common_attributes)}"
 
  resource_pool_id          = "${module.computer_cluster.producao_rj_resource_pool_id}"
 
  datastore_cluster_id      = "${module.datastore_cluster.producao_id}"

  num_cpus = "${var.num_cpus}"

  memory   = "${var.memory}"

  guest_id = "${data.vsphere_virtual_machine.template.guest_id}"

  scsi_type = "${data.vsphere_virtual_machine.template.scsi_type}"

  network_interface {
    network_id   = "${data.vsphere_network.network.id}"
  }

  disk {
    unit_number      = 0
    label            = "disk0"
    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
    eagerly_scrub    = "${data.vsphere_virtual_machine.template.disks.0.eagerly_scrub}"
    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
  }

  disk {
    unit_number      = 1
    label            = "disk1"
    size             = 50
    eagerly_scrub    = "${data.vsphere_virtual_machine.template.disks.0.eagerly_scrub}"
    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"

    customize {
      timeout = 0
      linux_options {
        host_name = "${var.host_name}"
        domain    = "globosat.net.br"
      }

      network_interface {
        ipv4_address = "${var.ipv4_address}"
        ipv4_netmask = 16
      }
      dns_server_list = ["10.1.0.12", "10.1.0.14"]
      dns_suffix_list = ["globosat.net.br"]
      ipv4_gateway = "10.9.0.1"
    }
  }
}
```
Thanks,

Marcelo Módolo

**Comment by teamterraform** [comment]
Hi @mrmodolo! 

Thank you for the additional information. You are running into a situation described [in the documentation](https://www.terraform.io/docs/configuration/resources.html#ignore_changes): 
 
> Adding or removing a key is treated by Terraform as a change to the containing map itself rather than to the individual key, and so if you wish to ignore changes to a particular tag made by an external system you must ensure that the Terraform configuration creates a placeholder element for that tag name so that the external system changes will be understood as an in-place edit of that key.

The shorter version of that caveat is: if you remove the key, there is no longer a key to check for changes. What you need to do is supply a dummy/placeholder value for `custom_attributes["103"]`, instead of removing it altogether, and terraform will ignore that value.

**Comment by mrmodolo** [comment]
Hi,

I had read the documentation but did not notice that it is not enough to just add the attribute with an empty value!

I tested the first time with the value below:
```
     "$ {module.attributes.last_backup_id}" = ""
```
After running with terraform apply the attribute seemed to continue to be ignored.

I then decided to test with a nonzero value.

```
     "$ {module.attributes.last_backup_id}" = "placehold"
```

After applying terraform apply and manually changing in the vmware console I no longer see terraform complain of changes!

```hcl
  lifecycle {
    ignore_changes = [
      boot_delay,
      guest_id,
      custom_attributes["103"],
    ]
  }
```
I have already tested several attribute changes by VMware and do not see the terraform command reporting changes.

So I believe it is not a bug, I lacked a better understanding of the documentation.

I apologize and appreciate the promptness!
